docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
+API. kubectl will use the discovery API to connect to these servers.
+
+These API servers will have full control of their resources.
+Each server will have their own storage mechanism and will implement their own
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
+
+  // Path to the swagger spec.
+  SwaggerSpecPath string
+}
+
+type ThirdPartyAPIServerStatus struct {
+  Conditions []ThirdPartyAPIServerCondition
+}
+
+type ThirdPartyAPIServerConditionType string
+const (
+        ThirdPartyAPIServerValid ThirdPartyAPIServerConditionType = "Valid"
+)
+
+type ThirdPartyAPIServerCondition struct {
+  Type ThirdPartyAPIServerConditionType
+  // Status of the condition, one of True, False or Unknown.
+  Status ConditionStatus
+  Reason string
+  Message string
+}
+```
+
+There are a few requirements that the third party API servers should meet
+(for ex: name of resources provided by each server should be unique).
+ThirdPartyAPIServerController will look for new servers and will validate them.
+It will set the valid condition in the status accordingly.
+Note that the server is validated only once so it is possible to change it after
+the validation is done. We expect that the server will be deleted and added
+again, whenever its functionality is changed.
+
+### kubectl
+
+Users will be able to use resources provided by third party servers seamlessly
+as if they were provided by the core server.
+Kubectl today uses a RESTMapper to determine the list of valid resources and
+operations. We will switch it to use the discovery API so that we can add and remove resources
```

Comment:
RESTMapper needs to be split into the following responsibilities in kubectl:

1. Turn a user entered "resource name" into a fully qualified resource,
group, api version (which can then be used with a generic client in
combination with auth info)
2. Turn a kind + api version + group from a file on disk into a resource,
group, and api version (to get a generic client and then post to)
3. Get a list of all "resources" that a user should see
4. Get swagger doc for a particular resource

On the server, its role is more limited, and the interface on that side
should be simpler as well.


On Mon, Nov 16, 2015 at 11:09 PM, Nikhil Jindal <notifications@github.com>
wrote:

> In docs/proposals/third-party-api-servers.md
> <https://github.com/kubernetes/kubernetes/pull/17191#discussion_r45020538>
> :
>
> > +```
> > +
> > +There are a few requirements that the third party API servers should meet
> > +(for ex: name of resources provided by each server should be unique).
> > +ThirdPartyAPIServerController will look for new servers and will validate them.
> > +It will set the valid condition in the status accordingly.
> > +Note that the server is validated only once so it is possible to change it after
> > +the validation is done. We expect that the server will be deleted and added
> > +again, whenever its functionality is changed.
> > +
> > +### kubectl
> > +
> > +Users will be able to use resources provided by third party servers seamlessly
> > +as if they were provided by the core server.
> > +Kubectl today uses a RESTMapper to determine the list of valid resources and
> > +operations. We will switch it to use the discovery API so that we can add and remove resources
>
> RESTMapper has a few more fields than what can be populated using
> discovery API. For ex: Codec and ObjectConverter.
>
> I think we should be able to move their uses to server?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17191/files#r45020538>.
>